src/spikeinterface/core/waveform_tools.py

```diff
@@ -397,6 +398,288 @@ def _waveform_extractor_chunk(segment_index, start_frame, end_frame, worker_ctx)
+        worker_ctx["shm"] = shm
+        worker_ctx["all_waveforms"] = all_waveforms
+
+    # prepare segment slices
+    segment_slices = []
+    for segment_index in range(recording.get_num_segments()):
+        s0 = np.searchsorted(spikes["segment_index"], segment_index)
+        s1 = np.searchsorted(spikes["segment_index"], segment_index + 1)
+        segment_slices.append((s0, s1))
+    worker_ctx["segment_slices"] = segment_slices
+
+    return worker_ctx
+
+
+# used by ChunkRecordingExecutor
+def _worker_ditribute_single_buffer(segment_index, start_frame, end_frame, worker_ctx):
+    # recover variables of the worker
+    recording = worker_ctx["recording"]
+    unit_ids = worker_ctx["unit_ids"]
+    segment_slices = worker_ctx["segment_slices"]
+    spikes = worker_ctx["spikes"]
+    nbefore = worker_ctx["nbefore"]
+    nafter = worker_ctx["nafter"]
+    return_scaled = worker_ctx["return_scaled"]
+    sparsity_mask = worker_ctx["sparsity_mask"]
+    all_waveforms = worker_ctx["all_waveforms"]
+
+    seg_size = recording.get_num_samples(segment_index=segment_index)
+
+    s0, s1 = segment_slices[segment_index]
+    in_seg_spikes = spikes[s0:s1]
+
+    # take only spikes in range [start_frame, end_frame]
+    # this is a slice so no copy!!
+    i0 = np.searchsorted(in_seg_spikes["sample_index"], start_frame)
+    i1 = np.searchsorted(in_seg_spikes["sample_index"], end_frame)
+    if i0 != i1:
+        # protect from spikes on border :  spike_time<0 or spike_time>seg_size
+        # useful only when max_spikes_per_unit is not None
+        # waveform will not be extracted and a zeros will be left in the memmap file
+        while (in_seg_spikes[i0]["sample_index"] - nbefore) < 0 and (i0 != i1):
+            i0 = i0 + 1
+        while (in_seg_spikes[i1 - 1]["sample_index"] + nafter) > seg_size and (i0 != i1):
+            i1 = i1 - 1
+
+    # slice in absolut in spikes vector
+    l0 = i0 + s0
+    l1 = i1 + s0
+
+    if l1 > l0:
+        start = spikes[l0]["sample_index"] - nbefore
+        end = spikes[l1 - 1]["sample_index"] + nafter
+
+        # load trace in memory
+        traces = recording.get_traces(
+            start_frame=start, end_frame=end, segment_index=segment_index, return_scaled=return_scaled
+        )
+
+        for spike_ind in range(l0, l1):
+            sample_index = spikes[spike_ind]["sample_index"]
+            unit_index = spikes[spike_ind]["unit_index"]
+            wf = traces[sample_index - start - nbefore : sample_index - start + nafter, :]
+
+            if sparsity_mask is None:
+                all_waveforms[spike_ind, :, :] = wf
+            else:
+                mask = sparsity_mask[unit_index, :]
+                wf = wf[:, mask]
+                all_waveforms[spike_ind, :, : wf.shape[1]] = wf
+
+        if worker_ctx["mode"] == "memmap":
+            all_waveforms.flush()
+
+
+def split_waveforms_by_units(unit_ids, spikes, all_waveforms, sparsity_mask=None, folder=None):
```

Comment:
Wheret this will be caled from?

Reply:
This is a future need: we will be able to replace in wavforms_extactors:
`extract_waveforms_to_buffers()`
by
`extract_waveforms_to_single_buffer()` + `split_waveforms_by_units()`
At the moment this is only used in the test to compare the two approach.

Reply:
Thanks for explaining.